### PR TITLE
SpanAttributes modified to use Vec instead of OrderMap/EvictedHashMap

### PR DESCRIPTION
**Pull request by cijothomas**
Towards https://github.com/open-telemetry/opentelemetry-rust/issues/1284 and https://github.com/open-telemetry/opentelemetry-rust/issues/1283. Only focused on Span attributes (to be extended to Events, Links post this PR).

This PR modifes Span attributes to use Vec<KeyValue> instead of OrderMap<Key, Value> in SpanBuilder( part of API), EvictedHashMap in Span (part of SDK). This is similar to what was done for [Logs](https://github.com/open-telemetry/opentelemetry-rust/pull/1142). 

Performance Implications:
The performance gains are matching what is already demonstrated in https://github.com/open-telemetry/opentelemetry-rust/pull/1121, but sharing the results from my run:

[Stress test](https://github.com/open-telemetry/opentelemetry-rust/blob/main/stress/src/traces.rs):
main branch:
Number of threads: 4
Throughput: 6,592,200 iterations/sec
Throughput: 6,504,400 iterations/sec
Throughput: 6,461,400 iterations/sec

This PR:
Number of threads: 4
Throughput: 10,274,200 iterations/sec
Throughput: 10,295,800 iterations/sec
Throughput: 10,329,000 iterations/sec

This is **~50% improvement** in throughput!

Benchmarks:
start-end-span-4-attrs/always-sample
	time:   [586.89 ns 594.65 ns 603.28 ns]
	**change: [-38.279%** -36.582% -34.871%] (p = 0.00 < 0.05)
	Performance has improved.

This means we'll no longer be doing de-duplicating of span attributes. [Opened an issue](https://github.com/open-telemetry/opentelemetry-rust/issues/1300) to track user feedback on this, which is also linked from changelog for easy visibility.

Additional Details: 
~The feature to enforce attributes limit is not done now. dropped_attribute_count will be 0 hence. It can be added back as a follow up, but I'll open a separate issue to discuss some issues in implementation of that, with some proposals.~ 

~Have just done the bare minimum to show the changes and its perf implications. I expect this to be 1st of a series of PRs doing similar for Events, Links, and also bringing back the attribute limit enforcement, along with new unit tests coverage. **_I want to get an okay that this direction is agreed upon, before working further on this_**. (Note: There were few other potential ideas discussed in todays SIG call, which can further help perf, but I'll open separate issues to track them., and they can be additive changes after this). Also, there might be interest in doing de-dup on a opt-in basis, either in SDK or a custom Processor, or Exporter  - this is another candidate for follow up!~

Thanks @shaun-cox and @lalitb for prior works, benchmarks and other consultation!

### CODE REVIEW

**Comment by shaun-cox** [comment]
I love the direction and think it's necessary to achieve the [project's mission](https://github.com/open-telemetry/community/blob/main/mission-vision-values.md#we-value-performance).

**Comment by cijothomas** [comment]
I made a quick attempt to fix [this ](https://github.com/open-telemetry/opentelemetry-rust/issues/1209) issue for Spans on top of this branch, and the throughput reaches 16m/sec. Overall that'd be more than doubling the perf (6M/sec to 16M/sec). Of course I haven't thought through the full implications, so will wait for green light on this PR's direction before further exploring that...

**Comment by jtescher** [comment]
@cijothomas I think this makes sense, thanks for working on it. Retaining first n rather than last n seems fine so the perf win here is well worth it. We probably still want to keep track of the dropped counts where possible, could store vec and droppped count, or have a struct that stored both?

**Comment by cijothomas** [comment]
> We probably still want to keep track of the dropped counts where possible, could store vec and droppped count, or have a struct that stored both

I can bring it back. I left some comments in here about that feature overall : https://github.com/open-telemetry/opentelemetry-rust/issues/1283#issuecomment-1757922468. I'd pursue that separately, to avoid distracting this PR.

Given there is +1s to the direction of this PR from other maintainer/approvers, I'll continue working on it.
Will mark it draft now, to polish it further and to add back the feature of tracking dropped_count. Once done, will undraf the PR to request full review! Thanks.

**Comment by TommyCpp** [comment]
I think it's the right direction. We should be explicit in docs that attributions can duplicate so we don't need to have deduplication or sort logic for them. Thanks for working on it

**Comment by jtescher** [comment]
could also see where perf is with hashmap to keep de-duplication but change to last-write-wins, would be worth knowing the cost of each

**Comment by cijothomas** [comment]
> could also see where perf is with hashmap to keep de-duplication but change to last-write-wins, would be worth knowing the cost of each

Yes thats what I tried first!. From my stress tests:
6.2 M/sec main branch

7.5 M/sec with regular HashMap (so sdk does dedup, but no eviction).

10 M/Sec with vec, where we removed de-duplication as well. 

(I do see a future where users will ask for de-duplication in the sdk level. We can do that as an opt-in SpanProcessor or even an Exporter feature, so as to no penalize every user.)

**Comment by cijothomas** [comment]
@open-telemetry/rust-approvers Please review!

**Review by TommyCpp — APPROVED** [review]

**Review by lalitb — APPROVED** [review]

File: opentelemetry-datadog/src/exporter/model/mod.rs
Comment (shaun-cox) [inline opentelemetry-datadog/src/exporter/model/mod.rs:287]: Should we address this before merge?

File: opentelemetry-sdk/src/trace/tracer.rs
Comment (shaun-cox) [inline opentelemetry-sdk/src/trace/tracer.rs:123]: I wonder if we even need this as `static` seeing as how it's pretty cheap to create an empty `Vec`.  In fact, `Vec::new` is a `const fn`.

**Review by shaun-cox — APPROVED** [review]

File: opentelemetry-datadog/src/exporter/model/mod.rs
Comment (cijothomas) [reply opentelemetry-datadog/src/exporter/model/mod.rs:287]: I don't think so. Opened issue to tackle this : https://github.com/open-telemetry/opentelemetry-rust/issues/1311

I'll re-bring the topic of creating contrib repo and move more components there, to make it easier to make changes in the main repo.

**Comment by TommyCpp** [comment]
Discussed with @cijothomas and we will follow up with the nits. Merging this for now

File: opentelemetry-sdk/src/trace/tracer.rs
Comment (cijothomas) [reply opentelemetry-sdk/src/trace/tracer.rs:123]: makes sense. Tacking this in the next PR continuing this direction!

**Pull request merged**
